### ISSUE TITLE
FPP canvas can fail in service workers

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/user.js#1813]
_Originally posted by @mik0l in https://github.com/arkenfox/user.js/issues/1716#issuecomment-1963013656_

Not always, but sometimes there is a leak in "Service".
 1.  Start browser and paste into the address bar https://abrahamjuliot.github.io/creepjs/tests/workers.html
 2.  If there is no leak, close browser and return to step 1.

---

_Originally posted by @Thorin-Oakenpants in https://github.com/arkenfox/user.js/issues/1716#issuecomment-1963033751_

OK, so I can confirm (nightly 125) - the result in SWers sometimes (I get it almost every time) is not randomized

@tomrittervg see STR ~~in previous comment

this is without FPP
![noFPP](https://github.com/arkenfox/user.js/assets/16656956/37904468-2c8a-4876-a6cb-bcafea6ac816)

some FPP examples

![diff1](https://github.com/arkenfox/user.js/assets/16656956/93cb49e5-e7ae-40f3-a3b1-7a64c271bc55)

![diff2](https://github.com/arkenfox/user.js/assets/16656956/b618ae7d-03ce-4eaa-800f-677729277636)

**Comment by Thorin-Oakenpants** [comment]
^ ping! @tomrittervg reminder .. next reminder in 3 more days :)

**Comment by tomrittervg** [comment]
I have it in my list, I'm in training the next couple days, so I'm not sure if I'll get to it before next reminder ;)

**Comment by Thorin-Oakenpants** [comment]
no worries, I'll stop pinging you now :) But this is a blocker for #1804 and unlocking the hundreds of millions of arkenfox users into FPP for testing and reporting 👀

**Comment by Thorin-Oakenpants** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1885471

**Comment by tomrittervg** [comment]
It looks like the window.open patches in https://bugzilla.mozilla.org/show_bug.cgi?id=1878716 will resolve this also

**Comment by Thorin-Oakenpants** [comment]
don't forget https://bugzilla.mozilla.org/show_bug.cgi?id=1889762

**Comment by Thorin-Oakenpants** [comment]
@tomrittervg did you know (at least FF125+) that if you have RFP enabled _and_ you have ETP Strict enabled - that when you relax RFP canvas extraction, FPP kicks in

here's an example showing RFP is on with RFP values for speech engines, media devices, audio context keys + values, etc .. with said conditions above (canvas exception, ETP blue shield .. it's strict)

![tada](https://github.com/arkenfox/user.js/assets/16656956/263927c7-dda7-4bfe-8278-b3b771a39a5f)

I wonder if this would happen in TB next ESR since it's already in PB mode (hence FPP by default) - cc @pospeselr

**Comment by tomrittervg** [comment]
Yeah... I think we should have a bug on file for this.  Canvas is just unusual because there's 4 behaviors to account for across the 2 different modes and then 2 ways of exempting a site that affect the modes differently.

**Comment by Thorin-Oakenpants** [comment]
gah .. still failing

**Comment by Thorin-Oakenpants** [comment]
weird - TZP shows FPP is used .. but that creepy test is showing no changes - it always shows `0e8c9761` which is my real value

**Comment by Thorin-Oakenpants** [comment]
> Yeah... I think we should have a bug on file for this. Canvas is just unusual because there's 4 behaviors to account for across the 2 different modes and then 2 ways of exempting a site that affect the modes differently.

https://gitlab.torproject.org/tpo/applications/tor-browser/-/issues/42556

**Comment by Thorin-Oakenpants** [comment]
AFAICT this issue is not resolved @tomrittervg

**Comment by Thorin-Oakenpants** [comment]
indeed ... ^ https://bugzilla.mozilla.org/show_bug.cgi?id=1885471#c6

**Comment by tomrittervg** [comment]
@Thorin-Oakenpants Is this resolved?

**Comment by Thorin-Oakenpants** [comment]
I haven't built my own tests, so I can only go on what I can understand from others - and I'm only testing with FPP off vs FPP on (not exceptions, or granularity, or RFPTargets)

using Nightly, PS everything 100% for all time is sanitized on close, not that it should matter

https://abrahamjuliot.github.io/creepjs/tests/workers.html
- set ETP standard and new session
   - canvas: `1dad120c` in window and all workers
   - new FF session: same result
   - so `1dad120c` is my expected canvas hash if not protected and the creepy test is static
- set ETP Strict and new session
  - new FF session 1: canvas: `ee892b80` across the board = protected
  - new FF session 2: canvas: `1dad120c` across the board = **no protection anywhere**
  - new FF session 3: canvas: `1dad120c` across the board = **no protection anywhere**
  - ^ why? I have no idea

repeat ETP standard - even the first session - canvas: `1dad120c` across the board = **no protection anywhere**


let's try https://canvasblocker.kkapsner.de/test/test.html
- set ETP standard and new session
   - `27537df6e459634249685b357e4f46cd3cc4fe35586521fb30da6cbe0f5df372` is my result for all tests
   - new FF session: same result
   - so this is my expected hash if not protected and canvasblocker test is static
- set ETP Strict and new session
  - new FF session 1: canvas: `3646c7c8aca44d77b253d7621b4e3923490f95a73edb037b5a2ff81bb2e4c5d8` across the board = protected
  - new FF session 2: canvas: `5c3094fe22ce88615deaeb063ff4374cbe4df91bfcb35becbe83f0086c91ac0c ` across the board = protected
  - new FF session 3: canvas: `f15bd25dc1c7e84c4f264a871a896e150f08faffbdebf34f85d132de49c361ff  ` across the board = protected
   - ^ seems to be working

ETP Strict - new session: canvasblocker = protected and in the same session creepy = not protected - IIUIC

So I would say, nope, not fixed = well fixed AFAICT in that service worker now always matches the rest on creepy - but not fixed in that we can get different results - TBH I only got `protected` in creepy once. Every attempt since in every session on any standard and strict, I get `1dad120c`

so, nope

**Comment by abrahamjuliot** [comment]
I'm late. It should run the same for all users. For the dedicated worker, I use a nested worker, or blob inside a blob.

**Comment by Thorin-Oakenpants** [comment]
thanks @abrahamjuliot, good to know you had a long holiday
 
it's been so long now I've forgotten everything about this issue - one day when i get time I will simply add my own tests

**Comment by Thorin-Oakenpants** [comment]
not that I think this is the case here, but just mentioning for all you budding canvas sleuths
- [1981505](https://bugzilla.mozilla.org/show_bug.cgi?id=1981505) - FPP Canvas uses RFP Canvas site exceptions
- this is by design

**Comment by Thorin-Oakenpants** [comment]
going to close this as FPP canvas is totally broken and there is nothing I can say or do about it

edit: this is not a regression - read the next comment

**Comment by Thorin-Oakenpants** [comment]
https://discuss.privacyguides.net/t/librewolf-firefox-based-browser/148/404

this is NOT a regression and info about it will be disclosed at some point - https://bugzilla.mozilla.org/show_bug.cgi?id=2013976 - tom is fully aware, and has been since the change was made, that it exposes `getImageData` (and it exposes bypasses on the other canvas methods)

**Comment by any1here** [comment]
> https://discuss.privacyguides.net/t/librewolf-firefox-based-browser/148/404
> 
> this is NOT a regression and info about it will be disclosed at some point - https://bugzilla.mozilla.org/show_bug.cgi?id=2013976 - tom is fully aware, and has been since the change was made, that it exposes `getImageData` (and it exposes bypasses on the other canvas methods)

I did not specifically mean this issue, but rather that just adding configs on top of Firefox can't fix everything when the underlying code is broken. That was poor wording on my part.

**Comment by Thorin-Oakenpants** [comment]
> when the underlying code is broken

I'm talking about the canvas part (not really following or testing webgl stuff yet since it's not an issue in Tor Browser)

You seem to think this is "broken" or a "regression" when it is in fact **_by design_**

Look, I'm not particularly happy about it myself, since it breaks the first rule of resistance (protect the real value), but that's a tradeoff - for example RFP protects devicePixelRatio and related values, but can still be calculated elsewhere - so it's effective to a degree but adversarial if a script bypasses it. Canvas randomizing is adversarial, period - you can't hide that it's altered. 

So whilst I don't agree with the changes, I do understand _why_ - namely that for FF users' threat model, this is effective at neutering canvas fingerprinting in most scripts - and has zero perf or display/rendering issues.

That said, it's a footgun (and a terrible look IMO, e.g. competence) and one day they will simply have to revert it all - in other words, a waste of time, mark my words. 

Your patch isn't protecting the real value, it's still exposed - I'm not going to say how. You've made a number of comments/claims that LW is superior with hardened FPing and that when you move to FPP with targets it's also hardened and superior over FPP default. There is no such thing - that's not how it works. I had this discussion with phoenix but the guy lost the plot. You are wasting your time - either you go RFP in which case uses should use MB IMO, or you go FPP and let mozilla control that as they add more protections (of course, as a reminder, it goes without saying that in MB/LW/FF you need to mask your IP or the FPing resistance is also a waste of time).

Not going to get into a discussion about it - tried that with phoenix. I strongly suggest you do not go the FPP + all targets with exceptions approach - the logic/reasons for doing so are flawed and  "snake oil" so to speak - I will blame the phoenix guy with his BS. If you stick with RFP then there's no difference to MB (when it has persistent mode) and if you go default FPP without fiddling with it, then you are no different to a few config changes in FF. Either way you are obsolete IMO. And if you do what I said not to, then you are wasting your time and grifting

[1] things like per site webgl etc - additional NEW things to add are good - not saying everything you do is bad. What about uplifting dFPI exception toggling via the urlbar? Put your efforts into FF - time to ditch LW for all, bury it :P

**Comment by any1here** [comment]
> You seem to think this is "broken" or a "regression" when it is in fact **_by design_**

So, it failing like you mentioned in your comment before you closed the issue is not broken but is meant to behave that way?

> You've made a number of comments/claims that LW is superior with hardened FPing

[https://bugzilla.mozilla.org/show\_bug.cgi?id=2013976](https://bugzilla.mozilla.org/show%5C_bug.cgi?id=2013976) isn't fixing this earlier an improvement over the current status in Firefox, which is all I ever mentioned regarding better FPP protection? I never claimed LW is the magnum opus and often talk about its weaknesses and often mention using MB/TR instead in our Matrix and issue tracker. I don't care how many users we have, I just want to provide the best I can within what the user base expects/tolerates.

> and that when you move to FPP with targets it's also hardened and superior over FPP default.

I personally would have switched to the [default targets with the webcompat fixes long ago](https://codeberg.org/librewolf/issues/issues/2121#issuecomment-8587185), but I can't just rule with an iron fist.

> few config changes in FF. Either way you are obsolete IMO.

Being obsolete was one of [my first comments](https://codeberg.org/librewolf/issues/issues/2735#issue-2781216) before I joined.

There is still a lot of low-hanging fruit that has been dormant for many years in the Bugzilla, mainly with regard to hardening the build, which I have recently started going through to assess how viable it is to add. As long as there is a user base happy to use it, I don't see the harm in maintaining a different flavor of Firefox.

> things like per site webgl etc - additional NEW things to add are good - not saying everything you do is bad. What about uplifting dFPI exception toggling via the urlbar?

Upstreaming the per-site WebGL permission is planned, but I am currently rewriting it so it integrates better with the other permissions.

> Put your efforts into FF - time to ditch LW for all, bury it :P

I never intended to stick around this long, considering I had never touched the Firefox codebase before. All I had originally planned on doing was fixing the failing build so the release could finally be made. Now I might as well stick around since the people involved in the project have been great, and I don't want them using a browser with slow updates again.

Either way, I still plan to upstream what’s worth upstreaming or what they would consider a worthwhile addition, so it being a side product of LibreWolf isn’t all that bad.

Edit:
> go FPP and let mozilla control that as they add more protections

Do you have any insider information when this might be happening? 👀

**Comment by Thorin-Oakenpants** [comment]
> So, it failing like you mentioned in your comment before you closed the issue is not broken but is meant to behave that way?

this issue was about service workers. at some point it was patched but something wasn't right (IIRC I could still reproduce a issue, maybe not the same issue, and from memory it was intermittent) so I left this open. I haven't checked in a long time, so IDK what the state of that is. TBH

When I discovered the change that you're talking about (the "regression" stuff). ... anyway, that's when I decided I really can't be bothered with this anymore
- > [me] going to close this as FPP canvas is totally broken and there is nothing I can say or do about it

tl;dr - we're all going to be obsolete - just bury LW now, arkenfox isn't far behind and good luck dealing with upstream patches and having the patience of a fucking saint

edited to remove personal thoughts now that it has been read

**Comment by mik0l** [comment]
Looks like it's fixed in v151

**Comment by Thorin-II** [comment]
at first glance, I can't reproduce the old issue either - tried a dozen times (on nightly 153)
- new FF session
- normal window = good
- new private window = good, close pb window
- repeat pb window test, each time is good and a new hash

**Comment by Thorin-II** [comment]
> Do you have any insider information when this might be happening?

@any1here
- https://bugzilla.mozilla.org/show_bug.cgi?id=2036786 phase 3
- https://bugzilla.mozilla.org/show_bug.cgi?id=1954197 phase 2

**Comment by any1here** [comment]
> > Do you have any insider information when this might be happening?
> 
> [@any1here](https://github.com/any1here)
> 
> * https://bugzilla.mozilla.org/show_bug.cgi?id=2036786 phase 3
> * https://bugzilla.mozilla.org/show_bug.cgi?id=1954197 phase 2

Since I can't message you anywhere else, I hope you don't mind me reusing this issue. What do you think about [Brave's approach](https://brave.com/privacy-updates/38-webgl-webgpu-fingerprinting-protections/) to protect the WebGL extensions, by adding a randomized _per session, per site (eTLD+1) and per storage area_ value from a list of fake extensions into the list of real values? Considering that this only protects against hashing `getSupportedExtensions`, since you can still get the actual supported values via `getExtension`.

**Comment by Thorin-II** [comment]
just to be clear, I _really_ don't want to discuss FPing

without getting into any specifics and speaking from a purist/theoretical viewpoint
- each FP "metric" (define a metric however you like) will have it's own set of circumstances/quirks - so there is no one answer
- when you protect a metric it should be _benign_ and _robust_
   - benign = real world, non-paradoxical - e.g. returning touch on a mac, returning a samsung renderer/vendor on a mac etc - pretty basic stuff, but it goes deeper than that
   - robust
     - you don't "leak" the real/inferred/different value, because if you do you create a paradox (paradoxes are not benign)
     - **_all methods_**: i.e all the ways you can get the information: passive/headers, js, inference (mac fonts on a mac), all the apis - pretty much the DOM, CSSOM .. in fact what I like to call the BOM (Browser Object Model)
        - e.g. look at all the ways the browser/app language (not to be confused with website language/navigator.lang*) can be detected via forms/controls, default fonts, and l10n
        - e.g. timezoneoffset (which is equivalency of timezonename), it's not enough to control the timezone itself, you also need to patch lastmodified timestamp strings
     - **_all sources_** - for both first and 3rd parties: top level document, iframes (sandboxed, not sandboxed etc), workers (web worker, shared, dedicated, service, from a blob, etc), via wasm
- all randomizing (in a default group - e.g. Brave +shields) can be detected and when detected becomes paradoxical/ non-benign
   - it is trivial to detect Brave Shields because of, ironically, specific randomization (randomizing techniques themselves contain entropy, and by that I mean _how_ each specific metric is randomized, not just that it _is_ randomized)
- anything not benign can and will be used against you
   - and being benign solves a lot of compat issues

I will point to this comment https://bugzilla.mozilla.org/show_bug.cgi?id=2050515#c11 as it contains links to some docs

As for Brave's implementation, I haven't specifically looked at it.
  
On randomizing (vs limiting)
- there are almost always ways to negate randomizing such as `perceptual hashing` + `locality-sensitive hashing`
- my stance laid out in my draft doc says it all (almost all, I plan to add some more info) - randomizing IMO is ultimately a waste of time
- but randomizing is also unavoidable (for now)
  - the only examples I can think of are canvas/webgl rendering or reads of said rendering
- sometimes you could get away with it
   - if it's very hard to infer/calculate/get via any other method _and_ you randomize within a set of limited (benign/real-world) values
   - ^ e.g. hardwareConcurrency or deviceMemory

**Comment by any1here** [comment]
> just to be clear, I _really_ don't want to discuss FPing

Understandable. I appreciate you nonetheless taking time to respond. I’ll not comment on the rest of your post.

**Comment by rusty-snake** [comment]
@any1here there is an `WebGLRandomization` RFPTarget. IDK what's the current state, the last time I checked it, it was basically useless because WebGL wasn't randomized when `EfficientCanvasRandomization` was enabled.
So don't use it and/or RFPTargets, just FYI.

**Comment by any1here** [comment]
> there is an `WebGLRandomization` RFPTarget. IDK what's the current state, the last time I checked it, it was basically useless because WebGL wasn't randomized when `EfficientCanvasRandomization` was enabled.

I submitted the fix to upstream half a year ago https://bugzilla.mozilla.org/show_bug.cgi?id=2013976 😅 

@rusty-snake Maybe I can get your opinion on my implementation for limiting the WebGL extensions https://librewolf.dev/librewolf/source/commit/65f7df19792a1950de4d8e3ecfc7595ed3028c09#diff-091991313e1f784339719ad98851d1ed9bed4e38.

This always reports all extensions as supported for all users and relies on `getExtension` to filter out the actually unsupported ones. I know this does not defeat iterating over `getExtension`, but it prevents just hashing `getSupportedExtensions` without breaking web‑compatibility, since I am not aware of any use cases where one uses the return value of `getSupportedExtensions` directly without also checking the return value of `getExtension`.
